### PR TITLE
Fix actionOptions not taking into account by the upload provider

### PR DESCRIPTION
**Pull request by petersg83**
Fixes https://github.com/strapi/strapi/issues/12809

It implements `uploadStream` wrapping for `actionOptions`.

Thank you @neilpoulin for catching the issue and spotting the cause !

### CODE REVIEW

File: packages/core/upload/server/services/provider.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ const { streamToBuffer } = require('../utils/file');
 
 module.exports = ({ strapi }) => ({
   async upload(file) {
-    if (isFunction(strapi.plugin('upload').provider.uploadStream)) {
+    if (isFunction(strapi.plugin('upload').provider.original.uploadStream)) {
```
Comment (neilpoulin) [inline packages/core/upload/server/services/provider.js:8]: ah, interesting way to handle the optional uploadStream! Thanks for tackling this.

File: packages/core/upload/server/bootstrap.js [hunk 2]
```diff
@@ -49,27 +49,35 @@ const createProvider = config => {
...
-    delete(file, options = actionOptions.delete) {
-      return providerInstance.delete(file, options);
-    },
+  if (!providerInstance.delete) {
+    throw new Error(`The upload provider "${providerName}" didn't implement the delete method.`);
+  }
+
+  if (!providerInstance.upload && !providerInstance.uploadStream) {
+    throw new Error(
+      `The upload provider "${providerName}" didn't implement the uploadStream nor the upload method.`
```
Comment (alexandrebodin) [inline packages/core/upload/server/bootstrap.js:58]: I would use doesn't instead of didn't in the messages

**Review by alexandrebodin — COMMENTED** [review]
Overall looks very good just a wording comment :)

**Comment by gu-stav** [comment]
After this PR is released https://github.com/strapi/strapi/pull/12838 can be merged.

**Comment by alexandrebodin** [comment]
> After this PR is released #12959 can be merged.

Any reason for waiting for the doc PR as we are just fixing something that already exists?

**Review by alexandrebodin — APPROVED** [review]
LGTM:

**Comment by birdzai** [comment]
So this is still an issue? I noticed the 4.1.6 release notes saying this was included in the fixes, but I'm unable to upload to specific Cloudinary folders.

**Comment by gu-stav** [comment]
@birdzai Apologies for the confusion. The change ended up in the changelog accidentally. This issue is not fixed yet and will be re-opened tomorrow, because we had test failures which we could not explain and need further investigation, unfortunately.

**Comment by gu-stav** [comment]
Re-opened as: https://github.com/strapi/strapi/pull/13018

**Pull request merged**